File: csharp/1096.brace-expansion-ii.cs
/*
 * @lc app=leetcode id=1096 lang=csharp
 *
 * [1096] Brace Expansion II
 */

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

// @lc code=start
public class Solution {

    private static void Merge(IList<HashSet<string>> all, HashSet<string> ret, StringBuilder sb, int level) {
        if (level == all.Count) {
            ret.Add(sb.ToString());
            return;
        }
        foreach (string s in all[level]) {
            sb.Append(s);
            Merge(all, ret, sb, level + 1);
            sb.Remove(sb.Length - s.Length, s.Length);
        }
    }

    private static HashSet<string> BraceExpansionIIRecurse(string expression, int n, ref int i) {
        var sb = new StringBuilder();
        if (expression[i] != '{') {
            while (i < n && Char.IsLetter(expression[i])) {
                sb.Append(expression[i++]);
            }
            return new HashSet<string>() { sb.ToString() };
        }
        ++i;
        var ret = new HashSet<string>();
        while (i < n && expression[i] != '}') {
            var all = new List<HashSet<string>>();
            while (i < n && expression[i] != ',' && expression[i] != '}') {
                all.Add(BraceExpansionIIRecurse(expression, n, ref i));
            }
            Merge(all, ret, sb, 0);
            if (i < n && expression[i] == ',') {
                ++i;
            }
        }
        ++i;
        return ret;
    }

    public IList<string> BraceExpansionII(string expression) {
        int i = 0;
        return BraceExpansionIIRecurse($"{{{expression}}}", 2 + expression.Length, ref i).OrderBy(s => s).ToList();
    }
}
// @lc code=end

File: java/1096.brace-expansion-ii.java
/*
 * @lc app=leetcode id=1096 lang=java
 *
 * [1096] Brace Expansion II
 */

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

// @lc code=start
class Solution {
    public List<String> braceExpansionII(String expression) {
        return helper("{%s}".formatted(expression), 0).results().stream().sorted().toList();
    }

    private Result helper(String expression, int index) {
        int n = expression.length(), i = index;
        if (expression.charAt(i) != '{') {
            while (i < n && expression.charAt(i) != ',' && expression.charAt(i) != '}'&& expression.charAt(i) != '{') {
                ++i;
            }
            return new Result(Set.of(expression.substring(index, i)), i);
        }
        ++i;
        Set<String> ret = new HashSet<>();
        while (i < n && expression.charAt(i) != '}') {
            List<Set<String>> sections = new ArrayList<>();
            while (i < n && expression.charAt(i) != '}' && expression.charAt(i) != ',') {
                Result sectionResult = helper(expression, i);
                sections.add(sectionResult.results());
                i = sectionResult.nextIndex();
            }
            combine(new StringBuilder(), ret, sections, 0);
            if (i < n && expression.charAt(i) == ',') {
                ++i;
            }
        }
        return new Result(ret, i + 1);
    }

    private static void combine(StringBuilder sb, Set<String> results, List<Set<String>> sections, int i) {
        if (i == sections.size()) {
            results.add(sb.toString());
            return;
        }
        for (String s : sections.get(i)) {
            sb.append(s);
            combine(sb, results, sections, i + 1);
            sb.delete(sb.length() - s.length(), sb.length());
        }
    }

    private record Result(Set<String> results, int nextIndex) {}
}
// @lc code=end

